### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by osterman**
## what
* Implement accepter/requester cross-account VPC peering 
* Support provisioning peering from a "root" account

## why
* Enable two different aws accounts to share resources

## references
* https://github.com/hashicorp/terraform/issues/12448#issuecomment-284400866

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -186,7 +186,7 @@
       same "printed page" as the copyright notice for easier
       identification within third-party archives.
 
-   Copyright [yyyy] [name of copyright owner]
+   Copyright 2017-2018 Cloud Posse, LLC
```
Comment (aknysh) [inline LICENSE]:
```suggestion
   Copyright 2019 Cloud Posse, LLC
```

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Compiled files
+*.tfstate
+*.tfstate.backup
+.terraform.tfstate.lock.info
+
+# Module directory
+.terraform/
+.idea
+terraform-aws-vpc-peering.iml
```
Comment (aknysh) [inline .gitignore]:
```suggestion
*.iml
```

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# Compiled files
+*.tfstate
+*.tfstate.backup
+.terraform.tfstate.lock.info
+
+# Module directory
+.terraform/
```
Comment (aknysh) [inline .gitignore]:
```suggestion
.terraform
```

File: README.yaml [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+related:
+  - name: "terraform-aws-vpc"
+    description: "Terraform Module that defines a VPC with public/private subnets across multiple AZs with Internet Gateways"
+    url: "https://github.com/cloudposse/terraform-aws-vpc"
+  - name: "terraform-aws-vpc-peering"
+    description: "Terraform module to create a peering connection between two VPCs in the same AWS account"
+    url: "https://github.com/cloudposse/terraform-aws-vpc"
+  - name: "terraform-aws-kops-vpc-peering"
+    description: "Terraform module to create a peering connection between two VPCs"
```
Comment (aknysh) [inline README.yaml]:
```suggestion
    description: "Terraform module to create a peering connection between a backing services VPC and a VPC created by Kops"
```

File: README.yaml [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  This module supports performing this action from a 3rd account (e.g. a "root" account) by specifying the roles to assume for each member account.
+  
+  **IMPORTANT:** AWS allows a cross-account VPC Peering Connection to be deleted from either the requester's or accepter's side. However, Terraform only allows the VPC Peering Connection to be deleted from the requester's side by removing the corresponding `aws_vpc_peering_connection` resource from your configuration. [Read more about this](https://www.terraform.io/docs/providers/aws/r/vpc_peering_accepter.html) on Terraform's documentation portal.
+
+# How to use this project
+usage: |-
+  ```hcl
+  module "vpc_peering" {
+    source           = "git::https://github.com/cloudposse/terraform-aws-vpc-peering-cross-account.git?ref=master"
+    namespace        = "cp"
```
Comment (aknysh) [inline README.yaml]:
```suggestion
    namespace        = "eg"
```

File: variables.tf [hunk 6]
```diff
@@ -0,0 +1,42 @@
+variable "enabled" {
+  default     = "true"
+  description = "Set to false to prevent the module from creating or accessing any resources"
+}
+
+variable "namespace" {
+  description = "Namespace (e.g. `cp` or `cloudposse`)"
```
Comment (aknysh) [inline variables.tf]:
```suggestion
  description = "Namespace (e.g. `eg` or `cp`)"
```

File: requester.tf [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+variable "requester_region" {
+  type        = "string"
+  description = "Requester AWS region"
+}
+
+variable "requester_vpc_id" {
+  type        = "string"
+  description = "Requestor VPC ID filter"
```
Comment (aknysh) [inline requester.tf]:
```suggestion
  description = "Requester VPC ID filter"
```

File: accepter.tf [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+  provider = "aws.accepter"
+  vpc_id   = "${local.accepter_vpc_id}"
+}
+
+locals {
+  accepter_subnet_ids       = "${distinct(sort(flatten(data.aws_subnet_ids.accepter.*.ids)))}"
+  accepter_subnet_ids_count = "${length(local.accepter_subnet_ids)}"
+  accepter_vpc_id           = "${join("", data.aws_vpc.accepter.*.id)}"
+  accepter_account_id       = "${join("", data.aws_caller_identity.accepter.*.account_id)}"
+  accepter_region           = "${join("", data.aws_region.accepter.*.id)}"
```
Comment (aknysh) [inline accepter.tf]:
```suggestion
  accepter_region           = "${join("", data.aws_region.accepter.*.name)}"
```

File: accepter.tf [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+  destination_cidr_block    = "${lookup(local.requester_cidr_block_associations[count.index % local.requester_cidr_block_associations_count], "cidr_block")}"
+  vpc_peering_connection_id = "${join("", aws_vpc_peering_connection.requester.*.id)}"
+  depends_on                = ["data.aws_route_table.accepter", "aws_vpc_peering_connection_accepter.accepter", "aws_vpc_peering_connection.requester"]
+}
+
+# Accepter's side of the connection.
+resource "aws_vpc_peering_connection_accepter" "accepter" {
+  count                     = "${local.count}"
+  provider                  = "aws.accepter"
+  vpc_peering_connection_id = "${aws_vpc_peering_connection.requester.id}"
```
Comment (aknysh) [inline accepter.tf]:
```suggestion
  vpc_peering_connection_id = "${join("", aws_vpc_peering_connection.requester.*.id)}"
```

File: accepter.tf (hunk 8 above)
Comment (osterman) [reply accepter.tf]: Wonder how this worked?

Comment (aknysh) [reply accepter.tf]: `id` is always there.
if they don't specify it in the doc, it's usually the same as `name`.
but `id` and `name` in some cases have different behavior: `id` will be known to TF only after the resource is 100% provisioned, while TF could know the `name` sooner - this was used to fix some `count of ... can't be computed` errors

**Comment by osterman** [comment]
Works...
```
Apply complete! Resources: 12 added, 0 changed, 0 destroyed.

Outputs:

accepter_accept_status = active
accepter_connection_id = pcx-0e53f266xxxxx
requester_accept_status = active
requester_connection_id = pcx-0e53f266xxxx

```

**Pull request merged**